java/java-psi-impl/src/com/intellij/psi/impl/source/tree/java/MethodReferenceResolver.java

```diff
@@ -81,9 +81,19 @@ protected MethodCandidateInfo createCandidateInfo(@NotNull final PsiMethod metho
                                                               final boolean staticProblem,
                                                               final boolean accessible,
                                                               final boolean varargs) {
+              PsiSubstitutor subst = substitutor;
               final PsiExpressionList argumentList = getArgumentList();
               final PsiType[] typeParameters = reference.getTypeParameters();
-              return new MethodCandidateInfo(method, substitutor, !accessible, staticProblem, argumentList, myCurrentFileContext,
+
+              if (typeParameters.length > 0) {
+                final PsiTypeParameter[] methodTypeParameters = method.getTypeParameters();
+                final int until = Math.min(methodTypeParameters.length, typeParameters.length);
+                for (int i = 0; i < until; i++) {
+                  subst = subst.put(methodTypeParameters[i], typeParameters[i]);
```

Comment:
type arguments are processed in `getSubstitutor()` method, initial substitutor corresponds to the substitutor which is defined by the call site, like we call `List#add` on a list of `String`s, then call site substitutor should map `E (of java.util.List) -> String`.

Reply:
I removed this code since it is no longer needed